client/wfm/src/service/WfmService.ts

```diff
@@ -0,0 +1,241 @@
+        return Promise.reject(new Error('No result exists for workflow ' + workorderId +
+          '. The workflow back topic can only be used for a workflow that has begun'));
+      }
+
+      // -1 is a special value for 'no next step'
+      result.nextStepIndex = _.max([result.nextStepIndex - 1, -1]) || -1; // _.max returns `number?`
+
+      return self.resultService.update(result).then(function() {
+        return {
+          workorder,
+          workflow,
+          result,
+          nextStepIndex: result.nextStepIndex,
+          step: result.nextStepIndex > -1 ? workflow.steps[result.nextStepIndex] : null
+        };
+      });
+    });
+  }
+
+  /**
+   * Completing a single step for a workorder.
+   */
+  public completeStep(parameters: CompleteStepParams) {
+    const workorderId = parameters.workorderId;
+    const stepCode = parameters.stepCode;
+    const submission = parameters.submission;
+    return this.userService.readUser().then(profileData => {
+      return this.workorderSummary(workorderId).then(summary => {
+        const workorder = summary.workorder;
+        const workflow = summary.workflow;
+        const workorderResult = summary.result;
+
+        if (!workorderResult) {
+          // No result exists, The workflow should have been started
+          return Promise.reject(new Error('No result exists for workorder ' + workorderId +
+            '. The workflow done topic can only be used for a workflow that has begun'));
+        }
+
+        const step = _.find(workflow.steps, s => s.code === stepCode);
+
+        // If there is no step, then this step submission is invalid.
+        if (!step) {
+          // No result exists, The workflow should have been started
+          return Promise.reject(new Error('Invalid step to assign completed data for workorder ' + workorderId +
+            ' and step code ' + stepCode));
+        }
+
+        // Got the workflow, now we can create the step result.
+        const stepResult = {
+          step,
+          submission,
+          status: STATUS.COMPLETE,
+          timestamp: new Date().getTime(),
+          submitter: profileData.id
+        };
+
+        // The result needs to be updated with the latest step results
+        workorderResult.stepResults = workorderResult.stepResults || {};
+        workorderResult.stepResults[step.code] = stepResult;
+        workorderResult.status = this.checkStatus(workorder, workflow, workorderResult);
+        workorderResult.nextStepIndex = this.executeStepReview(workflow.steps, workorderResult).nextStepIndex;
+
+        return this.resultService.update(workorderResult).then(function() {
+          return {
+            workorder,
+            workflow,
+            result: workorderResult,
+            nextStepIndex: workorderResult.nextStepIndex,
+            step: workorderResult.nextStepIndex > -1 ? workflow.steps[workorderResult.nextStepIndex] : workflow.steps[0]
```

Comment:
That looks terrible and may cause some issues (saw things crashing in this area)